cmd/ssh.go

```diff
@@ -244,3 +178,226 @@ func SSH(cmd *cobra.Command, args []string) error {
+	}
+
+	if err := executeSSH(tmpDstPath, dstPath, connectionInfo, privKeyPath); err != nil {
+		return fmt.Errorf("failed to extract source: %w", err)
+	}
+
+	ui.Infof("✅  Successfully copied source directory to remote host")
+
+	return nil
+}
+
+func createTempContextFile(execID string) (*os.File, error) {
+	name := fmt.Sprintf("uw-context-%s.tar.gz", execID)
+	tmpFile, err := os.CreateTemp(os.TempDir(), name)
+	if err != nil {
+		return nil, err
+	}
+	return tmpFile, nil
+}
+
+func executeSCP(srcPath, tmpDstPath, dstPath string, connectionInfo types.ConnectionInfo, privKeyPath string) error {
+	scpCommandArgs := []string{"-r",
+		"-o", "StrictHostKeyChecking=no",
+		"-o", "UserKnownHostsFile=/dev/null",
+	}
+	if privKeyPath != "" {
+		scpCommandArgs = append(scpCommandArgs, "-i", privKeyPath)
+	}
+	scpCommandArgs = append(scpCommandArgs, srcPath, fmt.Sprintf("%s@%s:%s", connectionInfo.User, connectionInfo.Host, tmpDstPath))
+
+	scpCommand := exec.Command("scp", scpCommandArgs...)
+
+	stdout := &bytes.Buffer{}
+	stderr := &bytes.Buffer{}
+
+	scpCommand.Stdout = stdout
+	scpCommand.Stderr = stderr
+
+	ui.Infof("🔄 Copying source to %q", dstPath)
+
+	if err := scpCommand.Run(); err != nil {
+		if exitError, ok := err.(*exec.ExitError); ok {
+			// Exited with non-zero exit code
+			if status, ok := exitError.Sys().(syscall.WaitStatus); ok {
+				if status.ExitStatus() == 255 {
+					ui.Infof("The remote host closed the connection.")
+					return fmt.Errorf("failed to copy source: %w", err)
+				}
+			}
+			ui.Infof("Failed to copy source directory to remote host: %s", stderr.String())
+			return err
+		}
+		return fmt.Errorf("scp command failed: %v", err)
+	}
+
+	return nil
+}
+
+func executeSSH(srcPath, dstPath string, connectionInfo types.ConnectionInfo, privKeyPath string) error {
+	sshCommand := exec.Command(
+		"ssh",
+		"-o", "StrictHostKeyChecking=no",
+		"-o", "UserKnownHostsFile=/dev/null",
+		"-i", privKeyPath,
+		fmt.Sprintf("%s@%s", connectionInfo.User, connectionInfo.Host),
+		fmt.Sprintf("tar -xzf %s -C %s && rm -rf %s", srcPath, dstPath, srcPath),
+	)
+
+	stdout := &bytes.Buffer{}
+	stderr := &bytes.Buffer{}
+
+	sshCommand.Stdout = stdout
+	sshCommand.Stderr = stderr
+
+	if err := sshCommand.Run(); err != nil {
+		if exitError, ok := err.(*exec.ExitError); ok {
+			// Exited with non-zero exit code
+			if status, ok := exitError.Sys().(syscall.WaitStatus); ok {
+				if status.ExitStatus() == 255 {
+					ui.Infof("The remote host closed the connection.")
+					return fmt.Errorf("failed to copy source: %w", err)
+				}
+			}
+			ui.Infof("Failed to extract source directory on remote host: %s", stderr.String())
+			return err
+		}
+		return fmt.Errorf("failed to unzip on remote host: %v", err)
+	}
+
+	return nil
+}
+
+// getPrivateKeyPathFromArgs returns the first best private key path from SSH arguments for an Exec
+func getPrivateKeyPathFromArgs(ctx context.Context, e types.Exec, sshArgsByKey map[string]string) string {
```

Comment:
I don't understand what this function does. I think too much modularization is also not good since it make code hard to ready

Reply:
Renamed and improved the code clarity, now it's `getUnweavePrivateKeyOrDefault`